packages/elements/src/react/common/loading.tsx

```diff
@@ -0,0 +1,249 @@
+  }
+
+  return true;
+}
+
+/**
+ * Access the loading state of a chosen scope. Scope can refer to a step, a provider, or the global loading state. The global loading state is `true` when any of the other scopes are loading.
+ *
+ * @param scope - Optional. Specify which loading state to access. Can be a step, a provider, or the global loading state. If `<Loading>` is used outside a `<Step>` it will default to `"global"`. If used inside a `<Step>` it will default to the current step.
+ * @param {Function} children - A render prop function that receives `isLoading` as an argument. `isLoading` is a boolean that indicates if the current scope is loading or not.
+ *
+ * @example
+ * <SignIn>
+ *   <Loading>
+ *     {(isLoading) => isLoading && "Global loading..."}
+ *   </Loading>
+ * </SignIn>
+ *
+ * @example
+ * <Step name="start">
+ *  <Submit>
+ *    <Loading>
+ *      {(isLoading) => isLoading ? "Start is loading..." : "Submit"}
+ *    </Loading>
+ *  </Submit>
+ * </Step>
+ *
+ * @example
+ * <Step name="start">
+ *   <Loading scope="provider:google">
+ *     {(isLoading) => (
+ *       <Provider name="google" disabled={isLoading}>
+ *         {isLoading ? "Loading..." : "Continue with Google"}
+ *       </Provider>
+ *     )}
+ *   </Loading>
+ * </Step>
+ */
+export function Loading({ children, scope }: LoadingProps) {
+  const signInRouterRef = SignInRouterCtx.useActorRef(true);
+  const signUpRouterRef = SignUpRouterCtx.useActorRef(true);
+
+  // One of the two routers must be defined. Otherwise, the component is used outside of a <SignIn> or <SignUp> component.
+  // An error is thrown in this case. Depending on which router is defined, the following flow will be either for sign-in or sign-up.
+  if (!signInRouterRef && !signUpRouterRef) {
+    throw new ClerkElementsRuntimeError(`<Loading> must be used within a <SignIn> or <SignUp> component.`);
+  }
+
+  if (signInRouterRef) {
+    if (isSignInScope(scope)) {
+      return (
+        <SignInLoading
+          scope={scope}
+          routerRef={signInRouterRef}
+        >
+          {children}
+        </SignInLoading>
+      );
+    } else {
+      throw new ClerkElementsRuntimeError(`Invalid scope "${scope}" used for <Loading> inside <SignIn>.`);
+    }
+  }
+
+  if (signUpRouterRef) {
+    if (isSignUpScope(scope)) {
+      return (
+        <SignUpLoading
+          scope={scope}
+          routerRef={signUpRouterRef}
+        >
+          {children}
+        </SignUpLoading>
+      );
+    } else {
+      throw new ClerkElementsRuntimeError(`Invalid scope "${scope}" used for <Loading> inside <SignUp>.`);
+    }
+  }
+
+  throw new ClerkElementsRuntimeError('Invalid state for <Loading>. Please open an issue.');
+}
+
+type SignInLoadingProps = {
+  scope?: LoadingScope<SignInStep>;
+  children: (isLoading: boolean) => React.ReactNode;
+  routerRef: ActorSignIn;
+};
+
+function SignInLoading({ children, scope, routerRef }: SignInLoadingProps) {
```

Comment:
There is _some_ shared logic between SignInLoading and SignUpLoading but I felt that trying to DRY things up would complicate things. I think as-is it's more readable albeit we have some duplicate lines.